main.go

```diff
@@ -0,0 +1,543 @@
+// task. Returns true if it successfully claimed a task (regardless of whether
+// the task ran successfully) otherwise false.
+func FindAndRunTask() bool {
+	// Write to the signed urls channel, to request signed urls back on
+	// channel c.
+	signedURLsRequestChan <- signedURLsResponseChan
+	// Read the result.
+	signedURLs := <-signedURLsResponseChan
+	taskFound := false
+	// Each of these signedURLs represent an underlying Azure queue, there
+	// are multiple of these so that we can support priority. For this
+	// reason the worker must poll the Azure queues in order they are
+	// given.
+	for _, urlPair := range signedURLs.Queues {
+		// try to grab a task using the url pair (url pair = poll url + delete
+		// url)
+		task, err := SignedURLPair(urlPair).Poll()
+		if err != nil {
+			// This can be any error at all occurs in queryAzureQueue that
+			// prevents us from claiming this task.  Log, and continue.
+			log.Println(err)
+			continue
+		}
+		// Now we found a task, run it, and then exit the loop. This is because
+		// the loop is in order of priority, most important first, so we will
+		// run the most important task we find, and then return, ignorning
+		// remaining urls for lower priority tasks that might still be left to
+		// loop through, since by the time we complete the first task, maybe
+		// higher priority jobs are waiting, so we need to poll afresh.
+		log.Println("Task found")
+		taskFound = true
+		// If there is one or more messages the worker must claim the tasks
+		// referenced in the messages, and delete the messages.
+		err = task.claim()
+		if err != nil {
+			log.Printf("WARN: Not able to claim task %v\n", task.TaskId)
+			log.Println(err)
+			continue
+		}
+		task.setReclaimTimer()
+		err = task.fetchTaskDefinition()
+		if err != nil {
+			log.Printf("WARN: Not able to fetch task definition for task %v\n", task.TaskId)
+			log.Println(err)
+			continue
+		}
+		err = task.validatePayload()
+		if err != nil {
+			log.Printf("WARN: Not able to validate task payload for task %v\n", task.TaskId)
+			log.Println(err)
+			continue
+		}
+		err = task.run()
+		if err != nil {
+			log.Printf("WARN: Not able to run task %v\n", task.TaskId)
+			log.Println(err)
+			continue
+		}
+		break
+	}
+	return taskFound
+}
+
+// Queries the given Azure Queue signed url pair (poll url/delete url) and
+// translates the Azure response into a Task object
+func (urlPair SignedURLPair) Poll() (*TaskRun, error) {
+	queueMessagesList := new(QueueMessagesList)
+	// To poll an Azure Queue the worker must do a `GET` request to the
+	// `signedPollUrl` from the object, representing the Azure queue. To
+	// receive multiple messages at once the parameter `&numofmessages=N`
+	// may be appended to `signedPollUrl`. The parameter `N` is the
+	// maximum number of messages desired, `N` can be up to 32.
+	// Since we can only process one task at a time, grab only one.
+	resp, err := http.Get(urlPair.SignedPollUrl + "&numofmessages=1")
```

Comment:
In general I haven't implemented them because:

* the worker tries to grab tasks every 1 second (and exponential backoff would slow this down)
* the task only gets deleted if all the http calls are successful, so in the case where an http call fails, a second later the same worker will have tried to claim it again

In other words, I think the whole process gets retried every second, so an exponential backoff won't really save us much, as all the http requests are pre task-run

The exception I guess is post flight http requests, once the task is in progress. Then there is a strong need for exponential backoff, so that compute time isn't wasted (e.g. after an hour of running a task, we don't want a failed http request to require that job starts from beginning again).

Do you agree with this choice? I could be oversimplifying, let me know what you think.